dali/core/mm/pinned_pool_test.cc

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cuda_runtime.h>
+#include <gtest/gtest.h>
+#include <random>
+#include <vector>
+#include "dali/core/mm/async_pool.h"
+#include "dali/core/dev_buffer.h"
+#include "dali/core/mm/mm_test_utils.h"
+#include "dali/core/cuda_stream.h"
+#include "rmm/mr/host/pinned_memory_resource.hpp"
+#include "dali/test/tensor_test_utils.h"
+
+namespace dali {
+namespace mm {
+namespace test {
+
+TEST(MMPinnedAlloc, StageCopy) {
+  test_pinned_resource upstream;
+  {
+    CUDAStream stream;
+    stream = CUDAStream::Create(true);
+    stream_view sv(stream);
+    async_pool_base<memory_kind::pinned> pool(&upstream);
+    std::mt19937_64 rng;
+    const int N = 1<<20;
+    vector<uint8_t> pattern(N), copy_back(N);
+    DeviceBuffer<uint8_t> dev_buf;
+    dev_buf.resize(N);
+    UniformRandomFill(pattern, rng, 0, 255);
+    void *mem1 = pool.allocate(N);
+    memcpy(mem1, pattern.data(), N);
+    CUDA_CALL(cudaMemcpyAsync(dev_buf, mem1, N, cudaMemcpyHostToDevice, stream));
+    pool.deallocate_async(mem1, N, sv);
+    void *mem2 = pool.allocate_async(N, sv);
+    EXPECT_EQ(mem1, mem2);
+    CUDA_CALL(cudaMemcpyAsync(copy_back.data(), dev_buf, N, cudaMemcpyDeviceToHost, stream));
+    CUDA_CALL(cudaStreamSynchronize(stream));
+    pool.deallocate_async(mem1, N, sv);
+    EXPECT_EQ(pattern, copy_back);
+  }
+  upstream.check_leaks();
+}
+
+TEST(MMPinnedAlloc, SyncAndSteal) {
+  test_pinned_resource upstream;
+  {
+    CUDAStream s1, s2;
+    s1 = CUDAStream::Create(true);
+    s2 = CUDAStream::Create(true);
+    stream_view sv1(s1), sv2(s2);
+    const int N = 1<<24;
+    async_pool_base<memory_kind::pinned> pool(&upstream, true);
+    void *mem1 = pool.allocate_async(N, sv1);
+    CUDA_CALL(cudaMemsetAsync(mem1, 0, N, s1));
+    pool.deallocate_async(mem1, N, sv1);
+    void *mem2 = pool.allocate_async(N, sv2);
```

Comment:
Why this allocate would case synchronization?
Because of the size? If so I would add a comment.

Reply:
I'll add the comment in the next PR if there are no more serious issues.

Reply:
I wonder how many times this can surprise the user. That his allocation won't happen immediately, but can sync on other stream (and random from the caller of the allocation point of view).

Reply:
Well, if the alternative is to synchronize the device (or all of them, as would be the case of pinned memory), then I'd say the user wouldn't notice any _negative_ impact. Also, it's similar to happens in plain malloc - either you allocate from process-local heap (fast) or issue a syscall to expand the heap (slower).

Reply:
But in this case you provide a soft promise to allocate from the pool without any unnecessary delay.